File: examples/key/new_managed_key_totp/new_managed_key_totp.ts
import {
  AccessControl,
  AccessControlTotp,
  AuthenticityClient,
  KeyClient,
  KeyProtectionLevel,
  KeyType,
  Managed,
  ManagedKeyParams,
  RecordClient,
  Signer,
} from '@bloock/sdk';

// initialize the Key Client
const keyClient = new KeyClient();
// initialize the Authenticity Client
const authenticityClient = new AuthenticityClient();
// initialize the Record Client
const recordClient = new RecordClient();
// initialize the protection key level
const keyProtection = KeyProtectionLevel.SOFTWARE;

// Generate a Rsa2048 key
const keyType = KeyType.Rsa2048;
const _key = await keyClient.newManagedKey(
  new ManagedKeyParams(keyProtection, keyType)
);

// Create a TOTP access control for the created key
const totp = await keyClient.setupTotpAccessControl(new Managed(_key));

console.log(totp.secretQr); // base64url encoded string with the TOTP setup QRCode. You can scan this QR with your TOTP application client.
console.log(totp.secret); // secret code that you can use instead of the QRCode to setup your TOTP application client.
console.log(totp.recoveryCodes); // recovery codes you need to save, so you could use in case you loose your access control.

const record = await recordClient.fromString('Hello world').build();

// How to sign using the TOTP access control
const accessCode = new AccessControlTotp('code digit'); // get the code from your TOTP application (Google Authenticator). Ej: 731 049
await authenticityClient.sign(
  record,
  new Signer(_key, undefined, new AccessControl(accessCode))
);

// Recover in case you lose your access control
const _totpRecovered = await keyClient.recoverTotpAccessControl(
  new Managed(_key),
  totp.recoveryCodes[0] // recovery code you saved previously
);

File: examples/key/new_managed_key_totp/new_managed_key_totp.py
from bloock.client.record import RecordClient
from bloock.client.key import KeyClient
from bloock.client.authenticity import AuthenticityClient
from bloock.entity.key.key_protection_level import KeyProtectionLevel
from bloock.entity.key.managed_key_params import ManagedKeyParams
from bloock.entity.key.key_type import KeyType
from bloock.entity.key.managed import Managed
from bloock.entity.key.access_control_totp import AccessControlTotp
from bloock.entity.authenticity.signer import Signer

# initialize the Key Client
key_client = KeyClient()
# initialize the Authenticity Client
authenticity_client = AuthenticityClient()
# initialize the Record Client
record_client = RecordClient()
# initialize the protection key level
protection = KeyProtectionLevel.SOFTWARE

# Generate a Rsa2048 key
key_type = KeyType.Rsa2048
key = key_client.new_managed_key(ManagedKeyParams(protection, key_type))

# Create a TOTP access control for the created key
totp = key_client.setup_totp_access_control(Managed(key))

# base64url encoded string with the TOTP setup QRCode.
# You can scan this QR with your TOTP application client.
print(totp.secret_qr)
# secret code that you can use instead of the QRCode to setup your TOTP application client.
print(totp.secret)
# recovery codes you need to save, so you could use in case you loose your access control.
print(totp.recovery_codes)

record = record_client.from_string("Hello world").build()

# How to sign using the TOTP access control
access_control = AccessControlTotp("code digit")
authenticity_client.sign(record, Signer(key, None, access_control))

# Recover in case you lose your access control
totpRecovered = key_client.recover_totp_access_control(Managed(key), totp.recovery_codes[0])

